server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2639,22 +2639,23 @@ private int getMaxDataVolumesSupported(UserVmVO vm) {
         return maxDataVolumesSupported.intValue();
     }
 
-    private Long getDeviceId(long vmId, Long deviceId) {
+    private Long getDeviceId(UserVmVO vm, Long deviceId) {
         // allocate deviceId
-        List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        int maxDataVolumesSupported = getMaxDataVolumesSupported(vm);
+        List<VolumeVO> vols = _volsDao.findByInstance(vm.getId());
         if (deviceId != null) {
-            if (deviceId.longValue() > 15 || deviceId.longValue() == 3) {
-                throw new RuntimeException("deviceId should be 1,2,4-15");
+            if (deviceId.longValue() > maxDataVolumesSupported || deviceId.longValue() == 3) {
+                throw new RuntimeException("deviceId should be 1,2,4-" + maxDataVolumesSupported);
             }
             for (VolumeVO vol : vols) {
                 if (vol.getDeviceId().equals(deviceId)) {
-                    throw new RuntimeException("deviceId " + deviceId + " is used by vm" + vmId);
+                    throw new RuntimeException("deviceId " + deviceId + " is used by vm" + vm.getId());
                 }
             }
         } else {
             // allocate deviceId here
             List<String> devIds = new ArrayList<String>();
-            for (int i = 1; i < 15; i++) {
+            for (int i = 1; i < maxDataVolumesSupported; i++) {
```

Comment:
@sureshanaparti If the condition really should be `i < maxDataVolumesSupported` (which would make the maximum device id returned by the method be `maxDataVolumesSupported - 1`), then the check + error message above
```
if (deviceId.longValue() <= 0 || deviceId.longValue() > maxDataVolumesSupported || deviceId.longValue() == 3) {
    throw new RuntimeException("deviceId should be 1,2,4-" + maxDataVolumesSupported);
```
need to be changed so as not to include the value of `maxDataVolumesSupported` itself.
Otherwise, when `maxDataVolumesSupported` has value `6` (for example), the method would not ever return `6` when parameter `deviceId` is specified as `null` but would return `6` when parameter `deviceId` is specified as `6` (assuming device id `6` is not already in use).  Also, the error message would state "deviceId should be 1,2,4-6" whenever parameter `deviceId` would be specified as an invalid value, which would not be correct (as `5` should be the highest valid device id).